### ISSUE TITLE
PMMR size calculation on restart can be wrong

### ISSUE THREAD

**Issue by ignopeverell** [mimblewimble/grin#371]
On restart, we compute the size of the PMMR by checking the size of the underlying storage file. However in case of rewinds, the storage isn't truncated, which can lead to a corruption if:

* We're on branch A that extends the PMMR size by 10 (for example)
* Branch B has more work, making us switch
* Branch B  extends the PMMR by only 5
* Node is restarted

**Comment by antiochp** [comment]
After our discussion - this makes sense to me now. The file has got out of sync with our rewound view of it before the restart...

**Comment by yeastplume** [comment]
Just on this, I think we should probably fix this as it may lead to very subtle and difficult-to-track errors. 

What do you think the best way to approach this is? Slamming it into the PMMR file would be counterproductive, but perhaps a metadata store file beside the PMMR file that just keeps track of the expected size (for now). We'd also need to make sure that updating this persisted size and updates/rewinds to the sum tree is an atomic transaction, so they don't get out of sync.. think this issue is slightly harder than it appears on the surface

**Comment by ignopeverell** [comment]
Yes, and there's the issue of pruning too (not activated yet). I don't think we can get perfect atomicity but we can try to limit the window with proper use of OS `sync`. I like the idea of the metadata file

**Comment by sesam** [comment]
I think we rather loose sync work on crash/restart, than have wrong data. I
think CAP relates to this. We would rather sacrifice persistence of some
ops, and maximize consistency and atomicity.

Suggestion: append only log of what height and file ops we want to do, then
do this op, then log success. With hashes and checks, so even disc
corruption can't wreck this. Restore then only reads back to data up to
last successfully committed height and PMMRs.
And maybe we already depend on some database that can supply this, so it
not even necessary to reinvent this db feature. Cockroachdb does
approximately this and then some.


tl;dr slow but sure, asynch, writeout to permament storage of chain sync
height and PMMRs, rarely (saving IOPS) but surely, to never restart into
wrong state.

On Dec 1, 2017 5:57 PM, "Ignotus Peverell" <notifications@github.com> wrote:

Yes, and there's the issue of pruning too (not activated yet). I don't
think we can get perfect atomicity but we can try to limit the window with
proper use of OS sync. I like the idea of the metadata file

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/mimblewimble/grin/issues/371#issuecomment-348563842>,
or mute the thread

**Comment by antiochp** [comment]
Is this something we want to tackle for testnet2?
This and #544 seem to be the two issue outstanding that can cause corrupted sumtrees and/or consensus failures due to sumtree differences.

**Comment by ignopeverell** [comment]
I've been thinking about this some more, it's actually very unlikely to occur. So definitely not as critical as #544, even though it would still be nice to have.

**Comment by yeastplume** [comment]
#735 should hopefeully address this.. will find out in testnet2 :D
